File: frontend/src/types/monitoring.ts
export type MonitoringSubject = {
  id: number;
  nome: string;
};

export type MonitoringClass = {
  id: number;
  nome: string;
  identificador?: string | null;
};

export type MonitoringAssignment = {
  subject_id: number;
  class_id: number;
};

export type MonitoringOptions = {
  subjects: MonitoringSubject[];
  classes: MonitoringClass[];
  assignments?: MonitoringAssignment[];
  lesson_types: string[];
};

export type MonitoringSession = {
  id: number;
  status: string;
};

export type MonitoringStatus = {
  status: string;
  session_id?: number | null;
  camera_status: string;
  websocket_clients: number;
  last_event?: string | null;
  error?: string | null;
  student_name: string;
  behavior: string;
  confidence: number;
  recognized_students_now?: number;
  timestamp?: string | null;
};

export type MonitoringStartPayload = {
  camera_id?: number | null;
  disciplina_id: number;
  turma_id: number;
  tipo_aula: string;
};

export type MonitoringRealtimeEvent = {
  type: string;
  session_id?: number | string | null;
  student_name?: string;
  behavior?: string;
  confidence?: number;
  faces_detected?: number;
  people_detected?: number;
  recognized_students_now?: number;
  camera_status?: string;
  websocket_status?: string;
  timestamp?: string;
  payload?: Record<string, unknown>;
};

export type CreateMonitoringSessionPayload = {
  subject_id: number;
  class_id: number;
  lesson_type: string;
};

File: backend/app/schemas/monitoring.py
from datetime import datetime

from pydantic import BaseModel, Field, model_validator


class MonitoringStatusResponse(BaseModel):
    status: str
    session_id: int | None = None
    camera_status: str = "idle"
    websocket_clients: int = 0
    last_event: str | None = None
    error: str | None = None
    student_name: str = "Aguardando identificação"
    behavior: str = "Aguardando"
    confidence: float = 0
    timestamp: datetime | None = None


class MonitoringSubject(BaseModel):
    id: int
    nome: str


class MonitoringClass(BaseModel):
    id: int
    nome: str
    identificador: str | None = None


class MonitoringOptionsResponse(BaseModel):
    subjects: list[MonitoringSubject]
    classes: list[MonitoringClass]
    lesson_types: list[str]


class MonitoringSessionCreate(BaseModel):
    subject_id: int
    class_id: int
    lesson_type: str = Field(default="Exposição", min_length=1, max_length=80)


class MonitoringSessionRead(BaseModel):
    id: int
    status: str


class MonitoringSessionClose(BaseModel):
    status: str = "encerrada"


class MonitoringSessionSummary(BaseModel):
    id: int
    status: str
    start_time: datetime | None = None
    end_time: datetime | None = None
    lesson_type: str | None = None


class MonitoringStartRequest(BaseModel):
    camera_id: int | None = None
    disciplina_id: int | None = None
    turma_id: int | None = None
    tipo_aula: str | None = None
    subject_id: int | None = None
    class_id: int | None = None
    lesson_type: str | None = None

    @model_validator(mode="after")
    def normalize_legacy_and_new_names(self) -> "MonitoringStartRequest":
        self.subject_id = self.subject_id or self.disciplina_id
        self.class_id = self.class_id or self.turma_id
        self.lesson_type = self.lesson_type or self.tipo_aula or "Exposição"
        return self

    def to_session_create(self) -> MonitoringSessionCreate:
        if self.subject_id is None or self.class_id is None:
            raise ValueError("subject_id/disciplina_id and class_id/turma_id are required")
        return MonitoringSessionCreate(
            subject_id=self.subject_id,
            class_id=self.class_id,
            lesson_type=self.lesson_type or "Exposição",
        )


class MonitoringStopRequest(BaseModel):
    status: str = "encerrada"
